base/reduce.jl

```diff
@@ -604,62 +615,32 @@ julia> prod(1:5; init = 1.0)
-function _fast(::typeof(max), x::AbstractFloat, y::AbstractFloat)
-    ifelse(isnan(x),
-        x,
-        ifelse(x > y, x, y))
-end
-
-function _fast(::typeof(min),x::AbstractFloat, y::AbstractFloat)
-    ifelse(isnan(x),
-        x,
-        ifelse(x < y, x, y))
-end
-
-isbadzero(::typeof(max), x::AbstractFloat) = (x == zero(x)) & signbit(x)
-isbadzero(::typeof(min), x::AbstractFloat) = (x == zero(x)) & !signbit(x)
-isbadzero(op, x) = false
-isgoodzero(::typeof(max), x) = isbadzero(min, x)
-isgoodzero(::typeof(min), x) = isbadzero(max, x)
-
-function mapreduce_impl(f, op::Union{typeof(max), typeof(min)},
-                        A::AbstractArrayOrBroadcasted, first::Int, last::Int)
-    a1 = @inbounds A[first]
-    v1 = mapreduce_first(f, op, a1)
-    v2 = v3 = v4 = v1
-    chunk_len = 256
-    start = first + 1
-    simdstop  = start + chunk_len - 4
-    while simdstop <= last - 3
-        @inbounds for i in start:4:simdstop
-            v1 = _fast(op, v1, f(A[i+0]))
-            v2 = _fast(op, v2, f(A[i+1]))
-            v3 = _fast(op, v3, f(A[i+2]))
-            v4 = _fast(op, v4, f(A[i+3]))
-        end
-        checkbounds(A, simdstop+3)
-        start += chunk_len
-        simdstop += chunk_len
-    end
-    v = op(op(v1,v2),op(v3,v4))
-    for i in start:last
-        @inbounds ai = A[i]
-        v = op(v, f(ai))
-    end
-
-    # enforce correct order of 0.0 and -0.0
-    # e.g. maximum([0.0, -0.0]) === 0.0
-    # should hold
-    if isbadzero(op, v)
-        for i in first:last
-            x = @inbounds A[i]
-            isgoodzero(op,x) && return x
-        end
-    end
-    return v
+# The idea behind the following functions is to transform `IEEEFloat` values to native `Signed` values that obey the same `min`/`max` behavior.
+# Namely, the float values must match the semantics of `<` except that `-0.0 < +0.0` and `NaN` is preferred over any other value.
+# In the case that both arguments are `NaN`, either value may be returned.
+# If we reinterpret(Signed,::IEEEFloat), the signed-integer ordering is:
+#   -0.0, ... negative values in ascending magnitude ..., -Inf, -NaN, 0.0, ... positive values in ascending magnitude ..., +Inf, +NaN
+# The desired ordering for min is:
+#   NaN (unspecified sign), -Inf, ... negative values in descending magnitude ..., -0.0, +0.0, ... positive values in ascending magnitude ..., +Inf
+# The desired ordering for max is:
+#   -Inf, ... negative values in descending magnitude ..., -0.0, +0.0, ... positive values in ascending magnitude ..., +Inf, NaN (unspecified sign)
+# Achieving this ordering requires two steps:
+#   1) flip the ordering of all values with a negative sign
+#   2) circularly-shift the values so that NaNs are together at the correct end of the spectrum
+# The following functions perform this transformation and reverse it.
+# Under this scheme, we place a total order on every value (including NaNs), although the ordering of NaNs is an implementation detail.
```

Comment:
I believe that which NaN is returned is, unlike in sorting, not specified in the documentation. Returning the first makes sense if it is just as easy, though in this case, I expect it is not.

Reply:
The IEEE Floating Point Standards Committee has decided that the most significant bit of a NaN cannot be discerned; the "sign bit" of a NaN must not play any role nor reflect payload content.  I initiated that discussion hoping to provide MPFR a reason to fix some unexpected interaction inside the string/print logic that flipped the state of the most significant bit of any NaN.

To use NaNs in a nonstandard way is now more gravel than roadblock.

Reply:
@JeffreySarnoff Could you provide a link to this decision?

Reply:
From the IEEE-754 standard, we have

> For an operation with quiet NaN inputs, other than maximum and minimum operations [maxNum and minNum as defined in the standard favor numbers over NaNs, unlike the maximum and minimum functions defined in Julia], if a floating-point result is to be delivered the result shall be a quiet NaN which should be one of the input NaNs.

I read this as: we should return one of the NaN inputs, but it is unspecified which.

Reply:
This PR chooses `sNaN < qNaN < -qNaN < -sNaN` (i.e., which end we choose from depends on `min` vs `max`), with further nuance based on the payload. Thus, what value is returned among multiple NaNs is dictated by the sign, qbit, and payload.

I can't look it up right now, but I believe our current implementations of `min` and `max` choose among NaNs based on sign and input position and our current `minimum` and `maximum` choose purely based on input position. The `min`/`max` methods being discussed under #45532 would change `min`/`max` to use the NaN semantics of either native machine `+` or `-` instead.

So the primary hazard I see for this method (and our existing implementations) is that it doesn't distinguish primarily based on signalingness, nor does it quiet a sNaN (although by adopting `+`/`-` semantics, one of the #45532 would likely be future-proof to such a requirement). But I don't believe either of those are required for these functions and they certainly aren't satisfied by the current versions.

Reply:
You read it correctly.  There is no "preferred" method.  Selecting the NaN with the larger payload as the one that continues offers better logics.
I have never owned hardware that uses that method.   Early on, I examined the NaN selection logic and tested payload propagation.  Those  choices are burned into processor design years before we can see one work.

The decision was taken over several days at the start of February 2021 using 
the IEEE-756 LISTSERVE.  Here is [the first message](https://listserv.ieee.org/cgi-bin/wa?A2=STDS-754;d95d7d2d.21&S=) and you may click on Next to page through. The last entry is from Mike Cowlishaw on Thursday, 4 Feb 2021 17:50:49.